app/pages/learn/01_tutorial/04_mastering-the-api/02_invoke/00_methodhandle.md

```diff
@@ -0,0 +1,531 @@
+```
+
+Let's create a method that transforms any `boolean` value by negating it:
+
+```java
+private static boolean negate(boolean original) {
+    return !original;
+}
+```
+
+Let's also create a method that increments any given integer value:
+
+```java
+private static int increment(int original) {
+    return ++original;
+}
+```
+
+Method handles for these transformation methods:
+
+```java
+MethodHandle negate = lookup.findStatic(Adapters.class, "negate", MethodType.methodType(boolean.class, boolean.class));
+MethodHandle increment = lookup.findStatic(Adapters.class, "increment", MethodType.methodType(int.class, int.class));
+```
+
+Now we can create a new method handle with transformations applied to two of the original arguments:
+
+```java
+MethodHandle withFilters = MethodHandles.filterArguments(originalMethodHandle, 0, increment, null, null, negate);
+withFilters.invoke(3, "abc", 5L, false); // outputs "4abc5true"
+```
+
+When we want to perform pre-processing of one or more arguments before the invocation of the original `MethodHandle`, 
+we can use `MethodHandles.foldArguments` and provide it with the method handle of any combiner method which will accept arguments starting at any preferred position.
+
+### Return value filtering
+Similarly to arguments, we can use an adapter that will apply transformations to the return value.
+
+Let's imagine a situation where we have a method that returns a `String`, and we would like to channel any returned value from this method into another method 
+that replaces character `d` with `m` and brings the resulting value to the uppercase.
+
+Here's the original method handle for the `getSomeString` method which always returns the value `"dummy"`:
+
+```java
+MethodHandle getSomeString = lookup.findStatic(Adapters.class, "getSomeString", MethodType.methodType(String.class));
+```
+
+Here's the `resultTransform` method that performs transformations:
+
+```java
+private static String resultTransform(String value) {
+    return value.replace('d', 'm').toUpperCase();
+}
+```
+
+Here is the method handle for our transformator method:
+
+```java
+MethodHandle resultTransform = lookup.findStatic(Adapters.class, "resultTransform", MethodType.methodType(String.class, String.class));
+```
+
+Finally, this is the combination of the two method handles where the result returned by `getSomeString` 
+method is then provided to the `resultTransform` method and modified accordingly:
+
+```java
+MethodHandle getSomeUppercaseString = MethodHandles.filterReturnValue(getSomeString, resultTransform);
+System.out.println(getSomeUppercaseString.invoke()); // outputs: "MUMMY"
+```
+
+<a id="conclusion">&nbsp;</a>
+## Conclusion
+In this tutorial, we have looked into the Method Handles mechanism and learned how to efficiently use it.
+We know now, that the method handles provide means for efficient method invocation, but this mechanism is not meant to replace the Reflection API.
```

Comment:
I think this might look nicer if it was reformulated to something like
```suggestion
We now know, that method handles provide means for efficient method invocation, but this mechanism is not meant to replace the Reflection API.
```